lib/core/buffer.hpp

```diff
@@ -41,39 +35,351 @@
+     *
+     * Unlike pass_metadata this doesn't remove the metadata from the @c from_buf
+     * and requires that the @c to_buf has a metadata object to be copied into.
+     *
+     * @param[in] from_frame_id The frame ID to copy the metadata from
+     * @param[in] to_buf The buffer to copy the metadata into
+     * @param[in] to_frame_id The frame ID in the @c to_buf to copy the metadata into
+     */
+    void copy_metadata(int from_frame_id, GenericBuffer* to_buf, int to_frame_id);
+
+    /// The number of frames kept by this object
+    int num_frames;
+
+    /// Should we shut down (stop returning new frames)?
+    bool shutdown_signal;
+
+    /// The list of consumer names registered to this buffer
+    std::map<std::string, StageInfo> consumers;
+
+    /// The list of producer names registered to this buffer
+    std::map<std::string, StageInfo> producers;
+
+    /// The name of the buffer for use in debug messages.
+    std::string buffer_name;
+
+    /// The type of the buffer for use in writing data.
+    std::string buffer_type;
+
+    /// The pool of info objects
+    metadataPool* metadata_pool;
+
+    /// Array of buffer info objects, for tracking information about each buffer.
+    std::vector<metadataContainer*> metadata;
+
+protected:
+    /// The main lock for frame state management
+    std::recursive_mutex mutex;
+
+    /// The condition variable for calls to @c wait_for_full_buffer
+    std::condition_variable_any full_cond;
+
+    /// The condition variable for calls to @c wait_for_empty_buffer
+    std::condition_variable_any empty_cond;
+
+    void private_copy_metadata(int dest_frame_id, GenericBuffer* src, int src_frame_id);
+};
+
+/**
+ * @brief A buffer to manage the signalling between stages when those
+ * stages want to communicate data using a ring buffer.
+ *
+ * This class is a little strange in that it doesn't actually allocate
+ * any memory!  It was created to connect GPU pipelines that need to
+ * communicate via a ring buffer in GPU memory.  This class knows how
+ * big the buffer is and controls read and write access for the
+ * consumers and producers, but does not actually own the memory.
+ *
+ * Producers and consumers can produce or consume different numbers of
+ * elements, so this is a helpful way of connecting stages that
+ * operate at different cadences.
+ *
+ * In the yaml config file, one of these can be created by setting the
+ * @c kotekan_buffer value to @c "ring".
+ *
+ * @conf ring_buffer_size: the number of elements in the ring buffer
+ * @conf metadata_pool The name of the metadata pool to associate with the buffer
+ */
+class RingBuffer : public GenericBuffer {
+public:
+    /**
+     * @brief Build a new RingBuffer for connecting kotekan stages
+     * that want to produce & consume data with different sizes or at
+     * different cadences.
+     *
+     * Currently, only a single producer and single consumer are supported!
```

Comment:
I don't think we need multiple producers (that's a really unlikely case with a complex ring like this), but multiple consumers are likely.  Based on your other comment, do you intend to add support in this PR, or finish this PR based on this limited case?

Reply:
I think I would like to get this PR in and then do multi-consumer in another (hopefully much smaller!) PR.